core/src/main/java/org/apache/calcite/rel/metadata/RelMdAllPredicates.java

```diff
@@ -113,19 +115,40 @@ public RelOptPredicateList getAllPredicates(TableScan scan, RelMetadataQuery mq)
   }
 
   /**
-   * Extract predicates for a project.
+   * Extracts predicates for a project.
    */
   public RelOptPredicateList getAllPredicates(Project project, RelMetadataQuery mq) {
     return mq.getAllPredicates(project.getInput());
   }
 
   /**
-   * Add the Filter condition to the list obtained from the input.
+   * Extracts predicates for a Filter.
    */
   public RelOptPredicateList getAllPredicates(Filter filter, RelMetadataQuery mq) {
-    final RelNode input = filter.getInput();
-    final RexBuilder rexBuilder = filter.getCluster().getRexBuilder();
-    final RexNode pred = filter.getCondition();
+    return getAllFilterPredicates(filter, mq, filter.getCondition());
+  }
+
+  /**
+   * Extracts predicates for a Calc.
+   */
+  public RelOptPredicateList getAllPredicates(Calc calc, RelMetadataQuery mq) {
+    final RexProgram rexProgram = calc.getProgram();
+    RexNode condition = null;
+    if (rexProgram.getCondition() != null) {
+      condition = rexProgram.expandLocalRef(rexProgram.getCondition());
+    } else {
+      return mq.getAllPredicates(calc.getInput());
+    }
+    return getAllFilterPredicates(calc, mq, condition);
+  }
```

Comment:
You can put `return getAllFilterPredicates(calc, mq, condition);` in the if conditiion.

Reply:
> You can put `return getAllFilterPredicates(calc, mq, condition);` in the if conditiion.

Thanks.